operators/src/upgrade/controller/reconciler.rs

```diff
@@ -0,0 +1,388 @@
+                assert!(p.is_none());
+                resource
+            }
+        }
+    }
+
+    /// Remove the resource from the operator
+    pub(crate) async fn remove(&self, name: String) -> Option<ResourceContext> {
+        let mut i = self.inventory.write().await;
+        let removed = i.remove(&name);
+        if let Some(removed) = removed {
+            info!(name =? removed.name_any(), "removed from inventory");
+            return Some(removed);
+        }
+        None
+    }
+}
+
+impl ResourceContext {
+    /// Called when putting our finalizer on top of the resource.
+    #[tracing::instrument(fields(name = ?ua.name_any()))]
+    pub(crate) async fn put_finalizer(ua: Arc<UpgradeAction>) -> Result<Action, Error> {
+        Ok(Action::await_change())
+    }
+
+    #[tracing::instrument(fields(name = ?resource.name_any()) skip(resource))]
+    pub(crate) async fn delete_finalizer(resource: ResourceContext) -> Result<Action, Error> {
+        let ctx = resource.ctx.clone();
+        ctx.remove(resource.name_any()).await;
+        Ok(Action::await_change())
+    }
+
+    /// Clone the inner value of this resource
+    fn inner(&self) -> Arc<UpgradeAction> {
+        self.inner.clone()
+    }
+
+    /// Construct an API handle for the resource
+    fn api(&self) -> Api<UpgradeAction> {
+        Api::namespaced(self.ctx.k8s.clone(), &self.namespace().unwrap())
+    }
+
+    async fn patch_status(&self, status: UpgradeActionStatus) -> Result<UpgradeAction, Error> {
+        let status = json!({ "status": status });
+
+        let ps = PatchParams::apply(UPGRADE_OPERATOR);
+
+        let o = self
+            .api()
+            .patch_status(&self.name_any(), &ps, &Patch::Merge(&status))
+            .await
+            .map_err(|source| Error::K8sClientError { source })?;
+
+        debug!(name = ?o.name_any(), old = ?self.status, new =?o.status, "status changed");
+
+        Ok(o)
+    }
+
+    async fn start(&self) -> Result<Action, Error> {
+        let _ = self
+            .patch_status(UpgradeActionStatus::new(
+                UpgradeState::NotStarted,
+                Utc::now(),
+                ComponentsState::with_state(UpgradePhase::Waiting).convert_into_hash(),
+            ))
+            .await?;
+        Ok(Action::await_change())
+    }
+
+    async fn updating(&self) -> Result<Action, Error> {
+        let _ = self
+            .patch_status(UpgradeActionStatus::new(
+                UpgradeState::Updating,
+                Utc::now(),
+                ComponentsState::with_state(UpgradePhase::Updating).convert_into_hash(),
+            ))
+            .await?;
+        let opts: Vec<(String, String)> = vec![("image.tag".to_string(), "2.0.0".to_string())];
+        let path = UpgradeOperatorConfig::get_config()
+            .helm_client()
+            .get_values()?;
+        UpgradeOperatorConfig::get_config()
+            .helm_client()
+            .upgrade(vec![path], opts)
```

Comment:
This returns an error, why are we not capturing this? Controller should not panic if an upgrade fails, it must be available to watch and serve other Upgrade requests (even though it will always be 1 resource always).